File: source/MikhailKhalizev.Processor.x86/src/BinToCSharp/Plugin/JmpCallLoopSimple.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using MikhailKhalizev.Processor.x86.Abstractions;
using MikhailKhalizev.Processor.x86.Abstractions.Memory;
using MikhailKhalizev.Processor.x86.Utils;
using SharpDisasm;
using SharpDisasm.Udis86;

namespace MikhailKhalizev.Processor.x86.BinToCSharp.Plugin
{
    public class JmpCallLoopSimple : PluginBase
    {
        /// <inheritdoc />
        public JmpCallLoopSimple(Engine engine)
            : base(engine)
        {
            Engine.InstructionDecoded += EngineOnInstructionDecoded;
        }

        private void EngineOnInstructionDecoded(object sender, Instruction cmd)
        {
            if (cmd.Operands.Count < 1)
                return;

            var op = cmd.Operands[0];
            Address toAddr = 0;

            if ((cmd.IsAnyJump || cmd.IsAnyLoop || cmd.IsCall) && (op.type == ud_type.UD_OP_JIMM))
            {
                switch (op.size)
                {
                    case 8: op.lval.sqword = op.lval.@sbyte; break;
                    case 16: op.lval.sqword = op.lval.sword; break;
                    case 32: op.lval.sqword = op.lval.sdword; break;
                    default: throw new NotImplementedException();
                }

                toAddr = cmd.End + (Address)op.lval.sqword;
            }
            else if ((cmd.IsCall || cmd.Mnemonic == ud_mnemonic_code.UD_Ijmp)
                    && !cmd.BrFar
                    && op.type == ud_type.UD_OP_MEM
                    && op.@base == ud_type.UD_NONE
                    && op.index == ud_type.UD_NONE
                    && op.offset != 0
                    // Support only flat model (without segments) to increase accuracy.
                    && Engine.Mode != ArchitectureMode.x86_16
                    && Engine.CsBase == 0
                    && Engine.DsBase == 0)
            {
                Address where;
                switch (op.offset)
                {
                    case 8: where = op.lval.@sbyte; break;
                    case 16: where = op.lval.sword; break;
                    case 32: where = op.lval.sdword; break;
                    default: throw new NotImplementedException();
                }


                switch (cmd.GetEffectiveSegmentOfOperand(cmd.Operands[0]))
                {
                    case ud_type.UD_R_CS:
                        where += Engine.CsBase;
                        break;
                    case ud_type.UD_R_DS:
                        where += Engine.DsBase;
                        break;
                    case ud_type.UD_R_SS:
                        return;
                    default:
                        throw new NotImplementedException();
                }

                switch (op.size)
                {
                    case 16:
                        toAddr = Engine.Memory.GetFixSize(where, 2).GetUInt16();
                        break;
                    case 32:
                        toAddr = Engine.Memory.GetFixSize(where, 4).GetUInt32();
                        break;
                    default:
                        throw new NotImplementedException();
                }

                toAddr += Engine.CsBase;


                if (toAddr == 0)
                    return;


                var os =
                    "Вызов '" +
                    AddressNameConverter.GetResultName(
                        toAddr,
                        false,
                        false /* лучше true, но комментарий в комментарие.. */) +
                    "'.";

                cmd.Comments.Add(os);
            }
            else
                return;


            Address eip = toAddr - Engine.CsBase;
            if (Engine.Mode == ArchitectureMode.x86_16)
                eip &= 0xffff;
            toAddr = Engine.CsBase + eip;


            var notSuppressed = !Engine.SuppressDecode.Contains(toAddr, false);

            if (notSuppressed)
                Engine.AddressesToDecode.Add(toAddr);

            if (cmd.IsCall || op.type != ud_type.UD_OP_JIMM)
            {
                if (notSuppressed)
                    Engine.NewDetectedMethods.Add(new DetectedMethod(toAddr)); // create if not exist.
            }
            else
            {
                Engine.jmp_to_known_addr.TryGetValue(new JumpsToKnownAddresses(cmd.Begin), out var actual);
                if (actual == null)
                {
                    actual = new JumpsToKnownAddresses(cmd.Begin);
                    actual.To = new SortedSet<Address>();
                    Engine.jmp_to_known_addr.Add(actual);
                }

                actual.To.Add(toAddr);
                cmd.write_cmd = on_cmd_write;
            }
        }

        private string on_cmd_write(Engine engine, DetectedMethod dm, int cmd_index, List<string> comments_in_current_func)
        {
            engine.jmp_to_known_addr.TryGetValue(new JumpsToKnownAddresses(dm.Instructions[cmd_index].Begin), out var cur_jmp);
            if (cur_jmp == null)
                throw new InvalidOperationException();

            var to = cur_jmp.To.Single();

            var func_suffix = "";
            if (!dm.Labels.Contains(to))
            {
                func_suffix = "_func";

                if (dm.Begin <= to && to < dm.End)
                    comments_in_current_func.Add("Адрес перехода делит инструкцию в этой функции пополам.");
            }

            return dm.Instructions[cmd_index].ToCodeString(func_suffix);
        }
    }
}

File: source/old-cpp/src/x86/bin_to_cxx/plugin/jmp_call_loop_simple.cpp
#include <sstream>
#include <exo/exception.hpp>
#include "x86/bin_to_cxx.hpp" // write_addr_with_check_known_definitions
#include "x86/bin_to_cxx/plugin_base.hpp"
#include "x86/bin_to_cxx/plugin/jmp_call_loop_simple.hpp"

namespace raw_program {
namespace x86 {
namespace plugin {


/** jmp call loop relative or
  * calld_abs(memd_a32(ds, 0x101be1c8)) */

jmp_call_loop_simple::jmp_call_loop_simple(bin_to_cxx & obj_)
: plugin_base(obj_)
{
    register_on_instruction_decoded( exo::util::delegate_unsafe<void (cmd_info &)>::create(this, &jmp_call_loop_simple::on_instruction_decoded) );
}

// +
void jmp_call_loop_simple::on_instruction_decoded(cmd_info & cmd)
{
    struct ud_operand & op = cmd.operand[0];
    addr_type to_addr;

    if ((cmd.is_any_jump || cmd.is_any_loop || cmd.is_call)
            && (op.type == UD_OP_JIMM))
    {
        switch (op.size)
        {
        case 8: op.lval.sqword = op.lval.sbyte; break;
        case 16: op.lval.sqword = op.lval.sword; break;
        case 32: op.lval.sqword = op.lval.sdword; break;
        default:
            throw exo::exception::not_implemented();
            break;
        }

        to_addr = static_cast<addr_type>(cmd.end + op.lval.sqword);
    }
    else if ((cmd.is_call || cmd.mnemonic == UD_Ijmp)
            && cmd.br_far == false
            && op.type == UD_OP_MEM
            && op.base == UD_NONE
            && op.index == UD_NONE
            && op.offset != 0)
    {
        addr_type where;
        switch (op.offset)
        {
        case 8: where = op.lval.sbyte; break;
        case 16: where = op.lval.sword; break;
        case 32: where = op.lval.sdword; break;
        default:
            throw exo::exception::not_implemented();
            break;
        }


        switch (cmd.get_effective_segment_of_operand(0))
        {
        case UD_R_CS:
            where += cs_base();
            break;
        case UD_R_DS:
            where += ds_base();
            break;
        case UD_R_SS:
            return;

        default:
            throw exo::exception::not_implemented();
            break;
        }


        switch (op.size)
        {
        case 16:
            to_addr = mem(where, 2).get<uint16_t>();
            break;
        case 32:
            to_addr = mem(where, 4).get<uint32_t>();
            break;

        default:
            throw exo::exception::not_implemented();
            break;
        }

        to_addr += cs_base();


        if (to_addr == 0)
            return;


        std::ostringstream os;
        os << std::hex << std::showbase;
        os << "Вызов '";
        write_addr_with_check_known_definitions(os, to_addr, true, false /* лучше true, но комментарий в комментарие.. */);
        os << "'.";
        cmd.comments.push_back(os.str());
    }
    else
        return;


    addr_type _eip = to_addr - cs_base();
    if (mode() == 16)
        _eip &= 0xffff;
    to_addr = cs_base() + _eip;


    bool not_suppressed = ! suppress_decode().contains(to_addr, false);

    if (not_suppressed)
        addr_to_decode().insert(to_addr);

    if (cmd.is_call || op.type != UD_OP_JIMM)
    {
        if (not_suppressed)
            new_detected_funcs().insert(detected_func(to_addr)); // create if not exist.
    }
    else
    {
        jmp_to_known_addr()[cmd.begin].to.insert(to_addr);
        cmd.write_cmd = decltype(cmd.write_cmd)::create(this, &jmp_call_loop_simple::on_cmd_write);
    }
}

// +
void jmp_call_loop_simple::on_cmd_write(std::ostream & os, detected_func & df, size_t cmd_index, std::vector<std::string> & comments_in_current_func)
{
    auto cur_jmp = jmp_to_known_addr().find(df.instr[cmd_index].first -> begin);
    if (cur_jmp == jmp_to_known_addr().end())
        throw exo::exception::logic_error();

    if (cur_jmp -> second.to.size() != 1)
        throw exo::exception::logic_error();

    addr_type to = *(cur_jmp -> second.to.begin());

    std::string func_suffix;
    if (df.labels.find(to) == df.labels.end())
    {
        func_suffix =  "_func";

        //if (df.begin <= to && to < df.end)  // внутренняя ошибка компилятора: в push_access_scope, в cp/pt.c:219
        if (df.begin.get() <= to.get() && to.get() < df.end.get())
            comments_in_current_func.push_back("Адрес перехода делит инструкцию в этой функции пополам.");
    }

    df.instr[cmd_index].first -> write_simple(os, func_suffix);
}


} /* namespace plugin */
} /* namespace x86 */
} /* namespace raw_program */
